spacy/cli/project/run.py

```diff
@@ -67,21 +71,62 @@ def project_run(
         when you want to turn over execution to the command, and capture=True
         when you want to run the command more like a function.
     """
+    if mult_group_mutex is None:
+        mult_group_mutex = Lock()
     config = load_project_config(project_dir, overrides=overrides)
     commands = {cmd["name"]: cmd for cmd in config.get("commands", [])}
     workflows = config.get("workflows", {})
+    max_parallel_processes = config.get("max_parallel_processes")
     validate_subcommand(list(commands.keys()), list(workflows.keys()), subcommand)
     if subcommand in workflows:
         msg.info(f"Running workflow '{subcommand}'")
-        for cmd in workflows[subcommand]:
-            project_run(
-                project_dir,
-                cmd,
-                overrides=overrides,
-                force=force,
-                dry=dry,
-                capture=capture,
-            )
+        for workflow_item in workflows[subcommand]:
+            if isinstance(workflow_item, str):
+                project_run(
+                    project_dir,
+                    workflow_item,
+                    overrides=overrides,
+                    force=force,
+                    dry=dry,
+                    capture=capture,
+                    mult_group_mutex=mult_group_mutex,
+                )
+            else:
+                assert isinstance(workflow_item, dict)
+                assert len(workflow_item) == 1
+                steps_list = workflow_item["parallel"]
+                assert isinstance(steps_list[0], str)
+                completion_queue = Queue(len(steps_list))
+                processes = [
+                    Process(
+                        target=project_run,
+                        args=(project_dir, cmd),
+                        kwargs={
+                            "overrides": overrides,
+                            "force": force,
+                            "dry": dry,
+                            "capture": capture,
+                            "mult_group_mutex": mult_group_mutex,
+                            "completion_queue": completion_queue,
+                        },
+                    )
+                    for cmd in steps_list
+                ]
+                num_processes = len(processes)
+                if (
+                    max_parallel_processes is not None
+                    and max_parallel_processes < num_processes
+                ):
+                    num_processes = max_parallel_processes
+                process_iterator = iter(processes)
+                for _ in range(num_processes):
+                    next(process_iterator).start()
+                for _ in range(len(steps_list)):
+                    completion_queue.get()
+                    next_process = next(process_iterator, None)
+                    if next_process is not None:
+                        next_process.start()
```

Comment:
Wouldn't it be simpler to use a `multiprocessing.Pool`, or even `concurrent.futures.ProcessPoolExecutor`? That way you just tell the library how many processes you want running at once, and it takes care of scheduling the individual tasks. Or is the idea that you specifically want to kill each process after it's finished one job, and spin up a new fresh process for the next one? I don't think `concurrent.futures` exposes that level of control, but with `multiprocessing.Pool` it's just a matter of passing `maxtasksperchild=1` to the constructor.

Also, if you're using multiprocessing, you should probably spend a few moments thinking about which [start method](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods) you want. On Windows and macOS the default is to re-invoke python from scratch, which is a bit heavy but safe, and there aren't really any other options. But on Linux, the default is to use `fork`, which is fast, *but* if the parent process is doing anything complex (in particular, using multithreading, either directly or via some library like BLAS), then `fork` can cause all kinds of nasty problems. `forkserver` or `spawn` are safer defaults.

Reply:
I've investigated the various ways of parallelising commands and [documented the results with code snippets](https://github.com/richardpaulhudson/spacy_projects_multiprocessing). if you follow the argumentation there, the current architecture with multiprocessing queues represents the most appropriate way of solving the problem, but changes are necessary to better manage error handling and logging and to ensure cross-platform consistency.

I'd be very grateful for feedback, especially on the specific issues set out in the last section.